compiler/src/dotty/tools/dotc/ast/Trees.scala

```diff
@@ -327,9 +327,52 @@ object Trees {
 
   extension (mdef: untpd.DefTree) def mods: untpd.Modifiers = mdef.rawMods
 
-  abstract class NamedDefTree[-T >: Untyped](implicit @constructorOnly src: SourceFile) extends NameTree[T] with DefTree[T] {
+  /** PackageDef | NamedDefTree */
+  sealed trait WithEndMarker:
+    self: Attachment.Container =>
+
+    import WithEndMarker.*
+
+    final def endSpan(using Context): Span =
+      self.getAttachment(EndIndex) match
+        case Some(end) =>
+          val realName = endName.stripModuleClassSuffix.lastPart
+          Span(end - realName.length, end)
+        case none => NoSpan
+
+    protected def endName(using Context): Name
+
+    final def withEndIndex(index: Int): self.type =
+      self.withAttachment(EndIndex, index)
+
+    final def withEndIndex(copyFrom: WithEndMarker): self.type =
+      copyFrom.endIndex.foreach(withEndIndex)
+      this
+
+    final def dropEndIndex(): self.type =
+      self.removeAttachment(EndIndex)
+      this
+
+    protected def endIndex: Option[Int] = self.getAttachment(EndIndex)
+
+  object WithEndMarker:
+    /** Property key for trees with an `end` marker */
+    private val EndIndex: Property.StickyKey[Int] = Property.StickyKey()
+
+  end WithEndMarker
+
+  abstract class NamedDefTree[-T >: Untyped](implicit @constructorOnly src: SourceFile)
+  extends NameTree[T] with DefTree[T] with WithEndMarker {
     type ThisTree[-T >: Untyped] <: NamedDefTree[T]
 
+    protected def endName(using Context) =
+      if name == nme.CONSTRUCTOR then nme.this_
```

Comment:
Do we actually need to separate srcName from endName? Sounds like this could also be the srcName of the constructor.

Reply:
thank you for the suggestion, this improves the situation actually